src/read/tokio.rs

```diff
@@ -0,0 +1,328 @@
+
+        let new_initial_header_start = w.stream_position().await?;
+        new_files.values_mut().try_for_each(|f| {
+            f.header_start = f.header_start.checked_add(new_initial_header_start).ok_or(
+                ZipError::InvalidArchive("new header start from merge would have been to large"),
+            )?;
+            f.central_header_start = 0;
+
+            if let Some(old_data_start) = f.data_start.take() {
+                let new_data_start = old_data_start.checked_add(new_initial_header_start).ok_or(
+                    ZipError::InvalidArchive("new data start from merge would have been to large"),
+                )?;
+                f.data_start.get_or_init(|| new_data_start);
+            }
+            Ok::<_, ZipError>(())
+        })?;
+
+        self.reader.rewind().await?;
+
+        let length_to_read = self.shared.dir_start;
+
+        let mut limited_raw = (&mut self.reader as &mut R).take(length_to_read);
+        tokio::io::copy(&mut limited_raw, &mut w).await?;
+
+        Ok(new_files)
+    }
+
+    async fn get_directory_info_zip32(
+        footer: &spec::CentralDirectoryEnd,
+        cde_start_pos: u64,
+    ) -> ZipResult<CentralDirectoryInfo> {
+        let archive_offset = cde_start_pos
+            .checked_sub(footer.central_directory_size as u64)
+            .and_then(|x| x.checked_sub(footer.central_directory_offset as u64))
+            .ok_or(ZipError::InvalidArchive(
+                "Invalid central directory size or offset",
+            ))?;
+        let directory_start = footer.central_directory_offset as u64 + archive_offset;
+        let number_of_files = footer.number_of_files_on_this_disk as usize;
+
+        Ok(CentralDirectoryInfo {
+            archive_offset,
+            directory_start,
+            number_of_files,
+            disk_number: footer.disk_number as u32,
+            disk_with_central_directory: footer.disk_with_central_directory as u32,
+        })
+    }
+
+    pub async fn get_directory_info_zip64(
+        reader: &mut R,
+        footer: &spec::CentralDirectoryEnd,
+        cde_start_pos: u64,
+    ) -> ZipResult<Vec<ZipResult<CentralDirectoryInfo>>> {
+        reader
+            .seek(tokio::io::SeekFrom::End(
+                -(20 + 22 + footer.zip_file_comment.len() as i64),
+            ))
+            .await?;
+
+        let locator64 = spec::Zip64CentralDirectoryEndLocator::parse(reader).await?;
+
+        let search_upper_bound = cde_start_pos
+            .checked_sub(60)
+            .ok_or(ZipError::InvalidArchive(
+                "File cannot contain ZIP64 central directory end",
+            ))?;
+
+        let search_results = spec::Zip64CentralDirectoryEnd::find_and_parse(
+            reader,
+            locator64.end_of_central_directory_offset,
+            search_upper_bound,
+        )
+        .await?;
+
+        let results: Vec<Result<CentralDirectoryInfo, ZipError>> = search_results.iter().map(|(footer64, archive_offset)| {
```

Comment:
Factor out lines 109-130 and their copies in the sync version to a shared method. It can be called `validate_zip64_footers`.